crates/sparrow-execution/src/lib.rs

```diff
@@ -8,254 +8,193 @@
-        ]);
-
-        let output_fields = Fields::from(vec![
-            Field::new("ab", DataType::Int64, true),
-            Field::new("abc", DataType::Int64, true),
-        ]);
-
-        let input_data = Arc::new(StructArray::new(
-            input_fields.clone(),
-            vec![
-                Arc::new(Int64Array::from(vec![0, 1, 2, 3])),
-                Arc::new(Int64Array::from(vec![4, 7, 10, 11])),
-                Arc::new(Int64Array::from(vec![Some(21), None, Some(387), Some(87)])),
-            ],
-            None,
-        ));
-        let time = Arc::new(TimestampNanosecondArray::from(vec![0, 1, 2, 3]));
-        let subsort = Arc::new(UInt64Array::from(vec![0, 1, 2, 3]));
-        let key_hash = Arc::new(UInt64Array::from(vec![0, 1, 2, 3]));
-        let input_batch = Batch::new_with_data(
-            input_data,
-            time,
-            subsort,
-            key_hash,
-            RowTime::from_timestamp_ns(3),
-        );
-        input_tx.send(input_batch).await.unwrap();
-        std::mem::drop(input_tx);
-
-        execute(
-            "hello".to_owned(),
-            DataType::Struct(input_fields),
-            input_rx,
-            DataType::Struct(output_fields),
-            output_tx,
-        )
-        .await
-        .unwrap();
+//! Execution of physical plans.
+
+use hashbrown::HashMap;
+use sparrow_physical::StepId;
+use sparrow_transforms::TransformPipeline;
+use std::sync::Arc;
+
+use error_stack::ResultExt;
+use itertools::Itertools;
+use sparrow_arrow::Batch;
+use sparrow_core::debug_println;
+use sparrow_scheduler::{PipelineInputs, WorkerPool};
+use uuid::Uuid;
+
+mod error;
+mod source_tasks;
+mod write_channel_pipeline;
+pub use error::*;
+
+use crate::source_tasks::SourceTasks;
+use crate::write_channel_pipeline::WriteChannelPipeline;
+
+pub struct PlanExecutor {
+    worker_pool: WorkerPool,
+    source_tasks: SourceTasks,
+}
 
-        let output = output_rx.recv().await.unwrap();
-        let output = output.data().unwrap().as_struct();
-        let ab = output.column_by_name("ab").unwrap();
-        let abc = output.column_by_name("abc").unwrap();
-        assert_eq!(ab.as_primitive(), &Int64Array::from(vec![4, 8, 12, 14]));
-        assert_eq!(
-            abc.as_primitive(),
-            &Int64Array::from(vec![Some(25), None, Some(399), Some(101)])
-        );
-    }
+const DEBUG_EXECUTOR_CREATION: bool = true;
 
-    /// Execute a physical plan.
-    pub async fn execute(
+impl PlanExecutor {
+    /// Create the plan executor for the given plan.
+    ///
+    /// This creates the worker threads but does not inject inputs, meaning
+    /// the threads won't have anything to do until `run` is called.
+    pub fn try_new(
         query_id: String,
-        input_type: DataType,
-        mut input: tokio::sync::mpsc::Receiver<Batch>,
-        output_type: DataType,
+        plan: sparrow_physical::Plan,
+        mut inputs: HashMap<Uuid, tokio::sync::mpsc::Receiver<Batch>>,
```

Comment:
Inputs are provided as a map from uuid to receiver. This will change with your PR to be UUID to `Arc<dyn Source>`.